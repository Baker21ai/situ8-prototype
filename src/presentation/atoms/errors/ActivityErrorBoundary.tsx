/**
 * ActivityErrorBoundary - Comprehensive error boundary for activity components
 * Handles different error scenarios with specialized recovery strategies
 */

import React, { Component, ReactNode, ErrorInfo } from 'react';
import { ErrorFallback } from './ErrorFallback';

interface Props {
  children: ReactNode;
  fallback?: ReactNode;
  onError?: (error: Error, errorInfo: ErrorInfo) => void;
  isolate?: boolean; // If true, prevents error propagation to parent boundaries
  context?: string; // Context for better error reporting
  enableRecovery?: boolean; // Enable retry and recovery features
  enableTelemetry?: boolean; // Enable error telemetry
}

interface State {
  hasError: boolean;
  error: Error | null;
  errorInfo: ErrorInfo | null;
  errorId: string | null;
  retryCount: number;
  isRecovering: boolean;
}

export class ActivityErrorBoundary extends Component<Props, State> {
  private retryTimeoutId: NodeJS.Timeout | null = null;
  private readonly maxRetries = 3;
  
  constructor(props: Props) {
    super(props);
    this.state = {
      hasError: false,
      error: null,
      errorInfo: null,
      errorId: null,
      retryCount: 0,
      isRecovering: false
    };
  }

  static getDerivedStateFromError(error: Error): Partial<State> {
    // Update state so the next render will show the fallback UI
    const errorId = `error_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
    return {
      hasError: true,
      error,
      errorId,
      isRecovering: false
    };
  }

  componentDidCatch(error: Error, errorInfo: ErrorInfo) {
    const { onError, context, enableTelemetry = true } = this.props;
    
    // Update state with error info
    this.setState({ errorInfo });
    
    // Call custom error handler
    if (onError) {
      onError(error, errorInfo);
    }

    // Log error for debugging
    console.group(`🚨 ActivityErrorBoundary Error ${context ? `(${context})` : ''}`);
    console.error('Error:', error);
    console.error('Error Info:', errorInfo);
    console.error('Component Stack:', errorInfo.componentStack);
    console.groupEnd();

    // Send telemetry if enabled
    if (enableTelemetry) {
      this.sendErrorTelemetry(error, errorInfo);
    }

    // Save error to local storage for debugging
    this.saveErrorToLocalStorage(error, errorInfo);
  }

  private sendErrorTelemetry = (error: Error, errorInfo: ErrorInfo) => {
    try {
      // In a real app, this would send to your error reporting service
      const telemetryData = {
        timestamp: new Date().toISOString(),
        error: {
          name: error.name,
          message: error.message,
          stack: error.stack
        },
        errorInfo: {
          componentStack: errorInfo.componentStack
        },
        context: this.props.context,
        userAgent: navigator.userAgent,
        url: window.location.href,
        userId: this.getCurrentUserId(), // Implement based on your auth system
        sessionId: this.getSessionId()
      };

      // Example: Send to error reporting service
      // errorReportingService.captureException(telemetryData);
      
      console.log('📊 Error telemetry data:', telemetryData);
    } catch (telemetryError) {
      console.error('Failed to send error telemetry:', telemetryError);
    }
  };

  private saveErrorToLocalStorage = (error: Error, errorInfo: ErrorInfo) => {
    try {
      const errorData = {
        timestamp: new Date().toISOString(),
        error: {
          name: error.name,
          message: error.message,
          stack: error.stack
        },
        errorInfo: {
          componentStack: errorInfo.componentStack
        },
        context: this.props.context,
        retryCount: this.state.retryCount
      };

      const existingErrors = JSON.parse(
        localStorage.getItem('situ8_activity_errors') || '[]'
      );
      
      // Keep only last 10 errors to prevent storage bloat
      const updatedErrors = [errorData, ...existingErrors.slice(0, 9)];
      
      localStorage.setItem('situ8_activity_errors', JSON.stringify(updatedErrors));
    } catch (storageError) {
      console.error('Failed to save error to localStorage:', storageError);
    }
  };

  private getCurrentUserId = (): string | null => {
    // Implement based on your authentication system
    // For now, return a placeholder
    return localStorage.getItem('current_user_id') || null;
  };

  private getSessionId = (): string => {
    let sessionId = sessionStorage.getItem('session_id');
    if (!sessionId) {
      sessionId = `session_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
      sessionStorage.setItem('session_id', sessionId);
    }
    return sessionId;
  };

  private handleRetry = () => {
    const { retryCount } = this.state;
    
    if (retryCount >= this.maxRetries) {
      console.warn('Maximum retry attempts reached');
      return;
    }

    this.setState({ isRecovering: true });

    // Clear any existing timeout
    if (this.retryTimeoutId) {
      clearTimeout(this.retryTimeoutId);
    }

    // Exponential backoff: 1s, 2s, 4s
    const delay = Math.pow(2, retryCount) * 1000;
    
    this.retryTimeoutId = setTimeout(() => {
      this.setState({
        hasError: false,
        error: null,
        errorInfo: null,
        errorId: null,
        retryCount: retryCount + 1,
        isRecovering: false
      });
    }, delay);
  };

  private handleReset = () => {
    // Clear any pending retries
    if (this.retryTimeoutId) {
      clearTimeout(this.retryTimeoutId);
    }

    // Reset to initial state
    this.setState({
      hasError: false,
      error: null,
      errorInfo: null,
      errorId: null,
      retryCount: 0,
      isRecovering: false
    });
  };

  private handleReportIssue = () => {
    const { error, errorInfo, errorId } = this.state;
    
    // Create issue report data
    const issueData = {
      errorId,
      timestamp: new Date().toISOString(),
      error: error?.toString(),
      componentStack: errorInfo?.componentStack,
      context: this.props.context,
      userAgent: navigator.userAgent,
      url: window.location.href
    };

    // In a real app, this would open your issue reporting system
    console.log('🐛 Issue report data:', issueData);
    
    // Example: Open issue reporting form
    // const issueUrl = `${config.issueReportingUrl}?data=${encodeURIComponent(JSON.stringify(issueData))}`;
    // window.open(issueUrl, '_blank');
    
    // For now, copy to clipboard
    navigator.clipboard.writeText(JSON.stringify(issueData, null, 2))
      .then(() => alert('Error details copied to clipboard'))
      .catch(() => console.error('Failed to copy error details'));
  };

  componentWillUnmount() {
    if (this.retryTimeoutId) {
      clearTimeout(this.retryTimeoutId);
    }
  }

  render() {
    const { hasError, error, errorInfo, isRecovering, retryCount } = this.state;
    const { children, fallback, enableRecovery = true, context = 'Activity Component' } = this.props;

    if (hasError && error) {
      // Use custom fallback if provided
      if (fallback && !enableRecovery) {
        return fallback;
      }

      // Use enhanced error fallback with recovery features
      return (
        <ErrorFallback
          error={error}
          errorInfo={errorInfo}
          context={context}
          onRetry={enableRecovery ? this.handleRetry : undefined}
          onReset={enableRecovery ? this.handleReset : undefined}
          onReportIssue={this.handleReportIssue}
          isRecovering={isRecovering}
          retryCount={retryCount}
          maxRetries={this.maxRetries}
          canRetry={enableRecovery && retryCount < this.maxRetries}
        />
      );
    }

    return children;
  }
}

// Convenience wrapper for common use cases
export const ActivityErrorBoundaryWrapper: React.FC<{
  children: ReactNode;
  context?: string;
  isolate?: boolean;
}> = ({ children, context, isolate = false }) => (
  <ActivityErrorBoundary
    context={context}
    isolate={isolate}
    enableRecovery={true}
    enableTelemetry={true}
  >
    {children}
  </ActivityErrorBoundary>
);

// Higher-order component for easy integration
export function withActivityErrorBoundary<P extends object>(
  Component: React.ComponentType<P>,
  context?: string
) {
  const WrappedComponent = (props: P) => (
    <ActivityErrorBoundary context={context || Component.displayName || Component.name}>
      <Component {...props} />
    </ActivityErrorBoundary>
  );

  WrappedComponent.displayName = `withActivityErrorBoundary(${Component.displayName || Component.name})`;
  
  return WrappedComponent;
}