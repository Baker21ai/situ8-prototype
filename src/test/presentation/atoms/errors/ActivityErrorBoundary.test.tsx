import { describe, it, expect, vi, beforeEach } from 'vitest';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import { ActivityErrorBoundary } from '../../../../presentation/atoms/errors/ActivityErrorBoundary';

// Mock console.error to avoid noise in tests
const mockConsoleError = vi.spyOn(console, 'error').mockImplementation(() => {});

// Component that throws an error for testing
const ThrowingComponent = ({ shouldThrow = false, errorMessage = 'Test error' }: { shouldThrow?: boolean; errorMessage?: string }) => {
  if (shouldThrow) {
    throw new Error(errorMessage);
  }
  return <div data-testid="working-component">Component is working</div>;
};

// Component that throws async error
const AsyncThrowingComponent = ({ shouldThrow = false }: { shouldThrow?: boolean }) => {
  React.useEffect(() => {
    if (shouldThrow) {
      throw new Error('Async error');
    }
  }, [shouldThrow]);
  
  return <div data-testid="async-component">Async component</div>;
};

describe('ActivityErrorBoundary', () => {
  beforeEach(() => {
    vi.clearAllMocks();
    mockConsoleError.mockClear();
  });

  describe('Normal operation', () => {
    it('should render children when no error occurs', () => {
      render(
        <ActivityErrorBoundary>
          <ThrowingComponent shouldThrow={false} />
        </ActivityErrorBoundary>
      );
      
      expect(screen.getByTestId('working-component')).toBeInTheDocument();
      expect(screen.getByText('Component is working')).toBeInTheDocument();
    });

    it('should render multiple children without errors', () => {
      render(
        <ActivityErrorBoundary>
          <div data-testid="child-1">Child 1</div>
          <div data-testid="child-2">Child 2</div>
          <ThrowingComponent shouldThrow={false} />
        </ActivityErrorBoundary>
      );
      
      expect(screen.getByTestId('child-1')).toBeInTheDocument();
      expect(screen.getByTestId('child-2')).toBeInTheDocument();
      expect(screen.getByTestId('working-component')).toBeInTheDocument();
    });
  });

  describe('Error handling', () => {
    it('should catch and display error fallback when child component throws', () => {
      render(
        <ActivityErrorBoundary>
          <ThrowingComponent shouldThrow={true} errorMessage="Test component error" />
        </ActivityErrorBoundary>
      );
      
      expect(screen.queryByTestId('working-component')).not.toBeInTheDocument();
      expect(screen.getByText(/Something went wrong with the activity component/)).toBeInTheDocument();
      expect(screen.getByText(/Test component error/)).toBeInTheDocument();
    });

    it('should display generic error message for unknown errors', () => {
      render(
        <ActivityErrorBoundary>
          <ThrowingComponent shouldThrow={true} errorMessage="" />
        </ActivityErrorBoundary>
      );
      
      expect(screen.getByText(/An unexpected error occurred/)).toBeInTheDocument();
    });

    it('should show retry button in error state', () => {
      render(
        <ActivityErrorBoundary>
          <ThrowingComponent shouldThrow={true} />
        </ActivityErrorBoundary>
      );
      
      expect(screen.getByText('Try Again')).toBeInTheDocument();
      expect(screen.getByRole('button', { name: 'Try Again' })).toBeInTheDocument();
    });

    it('should show refresh page button in error state', () => {
      render(
        <ActivityErrorBoundary>
          <ThrowingComponent shouldThrow={true} />
        </ActivityErrorBoundary>
      );
      
      expect(screen.getByText('Refresh Page')).toBeInTheDocument();
      expect(screen.getByRole('button', { name: 'Refresh Page' })).toBeInTheDocument();
    });
  });

  describe('Error recovery', () => {
    it('should recover when retry button is clicked', async () => {
      const user = userEvent.setup();
      
      const TestComponent = () => {
        const [shouldThrow, setShouldThrow] = React.useState(true);
        
        React.useEffect(() => {
          // Simulate error recovery after a delay
          const timer = setTimeout(() => {
            setShouldThrow(false);
          }, 100);
          
          return () => clearTimeout(timer);
        }, []);
        
        return (
          <ActivityErrorBoundary>
            <ThrowingComponent shouldThrow={shouldThrow} />
          </ActivityErrorBoundary>
        );
      };
      
      render(<TestComponent />);
      
      // Should show error initially
      expect(screen.getByText(/Something went wrong/)).toBeInTheDocument();
      
      const retryButton = screen.getByText('Try Again');
      await user.click(retryButton);
      
      // Should attempt to recover
      expect(retryButton).toBeInTheDocument();
    });

    it('should handle refresh page button click', async () => {
      const user = userEvent.setup();
      
      // Mock window.location.reload
      const mockReload = vi.fn();
      Object.defineProperty(window, 'location', {
        value: { reload: mockReload },
        writable: true
      });
      
      render(
        <ActivityErrorBoundary>
          <ThrowingComponent shouldThrow={true} />
        </ActivityErrorBoundary>
      );
      
      const refreshButton = screen.getByText('Refresh Page');
      await user.click(refreshButton);
      
      expect(mockReload).toHaveBeenCalled();
    });
  });

  describe('Error information', () => {
    it('should display error message when available', () => {
      const customErrorMessage = 'Custom activity error occurred';
      
      render(
        <ActivityErrorBoundary>
          <ThrowingComponent shouldThrow={true} errorMessage={customErrorMessage} />
        </ActivityErrorBoundary>
      );
      
      expect(screen.getByText(customErrorMessage)).toBeInTheDocument();
    });

    it('should handle Error objects properly', () => {
      const CustomError = () => {
        throw new Error('Detailed error message');
      };
      
      render(
        <ActivityErrorBoundary>
          <CustomError />
        </ActivityErrorBoundary>
      );
      
      expect(screen.getByText(/Detailed error message/)).toBeInTheDocument();
    });

    it('should handle non-Error objects thrown', () => {
      const StringThrowingComponent = () => {
        throw 'String error'; // eslint-disable-line no-throw-literal
      };
      
      render(
        <ActivityErrorBoundary>
          <StringThrowingComponent />
        </ActivityErrorBoundary>
      );
      
      expect(screen.getByText(/String error/)).toBeInTheDocument();
    });
  });

  describe('Development vs Production behavior', () => {
    it('should show detailed error information', () => {
      const detailedError = new Error('Detailed development error');
      detailedError.stack = 'Error stack trace\n  at Component\n  at ErrorBoundary';
      
      const DetailedErrorComponent = () => {
        throw detailedError;
      };
      
      render(
        <ActivityErrorBoundary>
          <DetailedErrorComponent />
        </ActivityErrorBoundary>
      );
      
      expect(screen.getByText(/Detailed development error/)).toBeInTheDocument();
    });
  });

  describe('Multiple error scenarios', () => {
    it('should handle multiple consecutive errors', () => {
      const MultiErrorComponent = ({ errorCount }: { errorCount: number }) => {
        if (errorCount > 0) {
          throw new Error(`Error number ${errorCount}`);
        }
        return <div>No errors</div>;
      };
      
      const { rerender } = render(
        <ActivityErrorBoundary>
          <MultiErrorComponent errorCount={1} />
        </ActivityErrorBoundary>
      );
      
      expect(screen.getByText(/Error number 1/)).toBeInTheDocument();
      
      rerender(
        <ActivityErrorBoundary>
          <MultiErrorComponent errorCount={2} />
        </ActivityErrorBoundary>
      );
      
      expect(screen.getByText(/Error number 2/)).toBeInTheDocument();
    });
  });

  describe('Nested error boundaries', () => {
    it('should handle nested error boundaries correctly', () => {
      render(
        <ActivityErrorBoundary>
          <div data-testid="outer-content">Outer content</div>
          <ActivityErrorBoundary>
            <ThrowingComponent shouldThrow={true} errorMessage="Inner error" />
          </ActivityErrorBoundary>
          <div data-testid="outer-content-2">More outer content</div>
        </ActivityErrorBoundary>
      );
      
      // Outer content should still be visible
      expect(screen.getByTestId('outer-content')).toBeInTheDocument();
      expect(screen.getByTestId('outer-content-2')).toBeInTheDocument();
      
      // Inner error should be caught by inner boundary
      expect(screen.getByText(/Inner error/)).toBeInTheDocument();
    });
  });

  describe('Component lifecycle errors', () => {
    it('should catch errors in component lifecycle methods', () => {
      const LifecycleErrorComponent = () => {
        React.useEffect(() => {
          throw new Error('Effect error');
        }, []);
        
        return <div>Component with effect error</div>;
      };
      
      // Note: useEffect errors are not caught by error boundaries in React
      // This test documents the current behavior
      render(
        <ActivityErrorBoundary>
          <LifecycleErrorComponent />
        </ActivityErrorBoundary>
      );
      
      // Component should render normally (useEffect errors aren't caught)
      expect(screen.getByText('Component with effect error')).toBeInTheDocument();
    });
  });

  describe('Accessibility', () => {
    it('should have proper ARIA attributes in error state', () => {
      render(
        <ActivityErrorBoundary>
          <ThrowingComponent shouldThrow={true} />
        </ActivityErrorBoundary>
      );
      
      const errorContainer = screen.getByText(/Something went wrong/).closest('div');
      expect(errorContainer).toHaveAttribute('role', 'alert');
    });

    it('should have accessible button labels', () => {
      render(
        <ActivityErrorBoundary>
          <ThrowingComponent shouldThrow={true} />
        </ActivityErrorBoundary>
      );
      
      expect(screen.getByRole('button', { name: 'Try Again' })).toBeInTheDocument();
      expect(screen.getByRole('button', { name: 'Refresh Page' })).toBeInTheDocument();
    });

    it('should be keyboard accessible', async () => {
      const user = userEvent.setup();
      
      render(
        <ActivityErrorBoundary>
          <ThrowingComponent shouldThrow={true} />
        </ActivityErrorBoundary>
      );
      
      const tryAgainButton = screen.getByRole('button', { name: 'Try Again' });
      const refreshButton = screen.getByRole('button', { name: 'Refresh Page' });
      
      // Should be able to focus buttons
      await user.tab();
      expect(tryAgainButton).toHaveFocus();
      
      await user.tab();
      expect(refreshButton).toHaveFocus();
    });
  });

  describe('Performance', () => {
    it('should not re-render unnecessarily when no error occurs', () => {
      const renderSpy = vi.fn();
      
      const SpyComponent = () => {
        renderSpy();
        return <div>Spy component</div>;
      };
      
      const { rerender } = render(
        <ActivityErrorBoundary>
          <SpyComponent />
        </ActivityErrorBoundary>
      );
      
      expect(renderSpy).toHaveBeenCalledTimes(1);
      
      // Re-render with same props
      rerender(
        <ActivityErrorBoundary>
          <SpyComponent />
        </ActivityErrorBoundary>
      );
      
      expect(renderSpy).toHaveBeenCalledTimes(2);
    });
  });

  describe('Edge cases', () => {
    it('should handle null children gracefully', () => {
      render(
        <ActivityErrorBoundary>
          {null}
        </ActivityErrorBoundary>
      );
      
      // Should render without errors
      expect(screen.queryByText(/Something went wrong/)).not.toBeInTheDocument();
    });

    it('should handle undefined children gracefully', () => {
      render(
        <ActivityErrorBoundary>
          {undefined}
        </ActivityErrorBoundary>
      );
      
      // Should render without errors
      expect(screen.queryByText(/Something went wrong/)).not.toBeInTheDocument();
    });

    it('should handle empty children array', () => {
      render(
        <ActivityErrorBoundary>
          {[]}
        </ActivityErrorBoundary>
      );
      
      // Should render without errors
      expect(screen.queryByText(/Something went wrong/)).not.toBeInTheDocument();
    });
  });

  describe('Error logging', () => {
    it('should log errors to console in development', () => {
      render(
        <ActivityErrorBoundary>
          <ThrowingComponent shouldThrow={true} errorMessage="Logged error" />
        </ActivityErrorBoundary>
      );
      
      expect(mockConsoleError).toHaveBeenCalled();
    });
  });

  describe('Snapshot tests', () => {
    it('should match snapshot for normal state', () => {
      const { container } = render(
        <ActivityErrorBoundary>
          <div>Normal content</div>
        </ActivityErrorBoundary>
      );
      
      expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot for error state', () => {
      const { container } = render(
        <ActivityErrorBoundary>
          <ThrowingComponent shouldThrow={true} errorMessage="Snapshot error" />
        </ActivityErrorBoundary>
      );
      
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
