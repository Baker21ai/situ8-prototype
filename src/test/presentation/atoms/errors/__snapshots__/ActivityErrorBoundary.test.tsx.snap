// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ActivityErrorBoundary > Snapshot tests > should match snapshot for error state 1`] = `
<div
  class="min-h-[200px] flex items-center justify-center p-4"
>
  <div
    class="text-card-foreground flex flex-col gap-6 rounded-xl border w-full max-w-2xl bg-orange-50 border-orange-200 shadow-lg"
    data-slot="card"
  >
    <div
      class="@container/card-header grid auto-rows-min grid-rows-[auto_auto] items-start gap-1.5 px-6 pt-6 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6 pb-3"
      data-slot="card-header"
    >
      <div
        class="flex items-start justify-between"
      >
        <div
          class="flex items-center gap-3"
        >
          <div
            class="p-2 rounded-full bg-orange-100 text-orange-800"
          >
            <svg
              class="lucide lucide-alert-triangle h-5 w-5 text-orange-800"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m21.73 18-8-14a2 2 0 0 0-3.48 0l-8 14A2 2 0 0 0 4 21h16a2 2 0 0 0 1.73-3Z"
              />
              <path
                d="M12 9v4"
              />
              <path
                d="M12 17h.01"
              />
            </svg>
          </div>
          <div>
            <h4
              class="text-lg text-orange-800"
              data-slot="card-title"
            >
              Something went wrong in 
              Activity Component
            </h4>
            <div
              class="flex items-center gap-2 mt-1"
            >
              <span
                class="inline-flex items-center justify-center rounded-md border px-2 py-0.5 text-xs font-medium w-fit whitespace-nowrap shrink-0 [&>svg]:size-3 gap-1 [&>svg]:pointer-events-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive transition-[color,box-shadow] overflow-hidden [a&]:hover:bg-accent [a&]:hover:text-accent-foreground bg-orange-100 text-orange-800"
                data-slot="badge"
              >
                MEDIUM
                 SEVERITY
              </span>
            </div>
          </div>
        </div>
        <div
          class="flex items-center gap-1"
        >
          <button
            class="inline-flex items-center justify-center whitespace-nowrap font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:bg-accent hover:text-accent-foreground dark:hover:bg-accent/50 h-8 rounded-md gap-1.5 px-3 has-[>svg]:px-2.5 text-xs"
            data-slot="button"
          >
            <svg
              class="lucide lucide-copy h-3 w-3 mr-1"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                height="14"
                rx="2"
                ry="2"
                width="14"
                x="8"
                y="8"
              />
              <path
                d="M4 16c-1.1 0-2-.9-2-2V4c0-1.1.9-2 2-2h10c1.1 0 2 .9 2 2"
              />
            </svg>
            Copy
          </button>
          <button
            class="inline-flex items-center justify-center whitespace-nowrap font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:bg-accent hover:text-accent-foreground dark:hover:bg-accent/50 h-8 rounded-md gap-1.5 px-3 has-[>svg]:px-2.5 text-xs"
            data-slot="button"
          >
            <svg
              class="lucide lucide-bug h-3 w-3 mr-1"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m8 2 1.88 1.88"
              />
              <path
                d="M14.12 3.88 16 2"
              />
              <path
                d="M9 7.13v-1a3.003 3.003 0 1 1 6 0v1"
              />
              <path
                d="M12 20c-3.3 0-6-2.7-6-6v-3a4 4 0 0 1 4-4h4a4 4 0 0 1 4 4v3c0 3.3-2.7 6-6 6"
              />
              <path
                d="M12 20v-9"
              />
              <path
                d="M6.53 9C4.6 8.8 3 7.1 3 5"
              />
              <path
                d="M6 13H2"
              />
              <path
                d="M3 21c0-2.1 1.7-3.9 3.8-4"
              />
              <path
                d="M20.97 5c0 2.1-1.6 3.8-3.5 4"
              />
              <path
                d="M22 13h-4"
              />
              <path
                d="M17.2 17c2.1.1 3.8 1.9 3.8 4"
              />
            </svg>
            Report
          </button>
        </div>
      </div>
    </div>
    <div
      class="px-6 [&:last-child]:pb-6 space-y-4"
      data-slot="card-content"
    >
      <div
        class="relative w-full rounded-lg px-4 py-3 text-sm grid has-[>svg]:grid-cols-[calc(var(--spacing)*4)_1fr] grid-cols-[0_1fr] has-[>svg]:gap-x-3 gap-y-0.5 items-start [&>svg]:size-4 [&>svg]:translate-y-0.5 [&>svg]:text-current text-card-foreground border-0 bg-white/50"
        data-slot="alert"
        role="alert"
      >
        <div
          class="col-start-2 grid justify-items-start gap-1 text-sm [&_p]:leading-relaxed font-medium text-orange-800"
          data-slot="alert-description"
        >
          Snapshot error
        </div>
      </div>
      <div
        class="flex flex-wrap gap-2"
      >
        <button
          class="justify-center whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-primary text-primary-foreground hover:bg-primary/90 h-9 px-4 py-2 has-[>svg]:px-3 flex items-center gap-2"
          data-slot="button"
        >
          <svg
            class="lucide lucide-refresh-cw h-4 w-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3 12a9 9 0 0 1 9-9 9.75 9.75 0 0 1 6.74 2.74L21 8"
            />
            <path
              d="M21 3v5h-5"
            />
            <path
              d="M21 12a9 9 0 0 1-9 9 9.75 9.75 0 0 1-6.74-2.74L3 16"
            />
            <path
              d="M8 16H3v5"
            />
          </svg>
          Retry (
          3
           left)
        </button>
        <button
          class="justify-center whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive border bg-background text-foreground hover:bg-accent hover:text-accent-foreground dark:bg-input/30 dark:border-input dark:hover:bg-input/50 h-9 px-4 py-2 has-[>svg]:px-3 flex items-center gap-2"
          data-slot="button"
        >
          <svg
            class="lucide lucide-rotate-ccw h-4 w-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3 12a9 9 0 1 0 9-9 9.75 9.75 0 0 0-6.74 2.74L3 8"
            />
            <path
              d="M3 3v5h5"
            />
          </svg>
          Reset Component
        </button>
        <button
          class="justify-center whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive border bg-background text-foreground hover:bg-accent hover:text-accent-foreground dark:bg-input/30 dark:border-input dark:hover:bg-input/50 h-9 px-4 py-2 has-[>svg]:px-3 flex items-center gap-2"
          data-slot="button"
        >
          <svg
            class="lucide lucide-external-link h-4 w-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6"
            />
            <polyline
              points="15 3 21 3 21 9"
            />
            <line
              x1="10"
              x2="21"
              y1="14"
              y2="3"
            />
          </svg>
          Refresh Page
        </button>
      </div>
      <div
        class="space-y-2"
      >
        <h4
          class="font-medium text-sm text-orange-800"
        >
          Try these solutions:
        </h4>
        <ul
          class="space-y-1"
        >
          <li
            class="text-sm text-orange-800 flex items-start gap-2"
          >
            <span
              class="text-xs mt-1"
            >
              •
            </span>
            Try refreshing the page
          </li>
          <li
            class="text-sm text-orange-800 flex items-start gap-2"
          >
            <span
              class="text-xs mt-1"
            >
              •
            </span>
            Clear your browser cache
          </li>
          <li
            class="text-sm text-orange-800 flex items-start gap-2"
          >
            <span
              class="text-xs mt-1"
            >
              •
            </span>
            Contact support if the issue persists
          </li>
        </ul>
      </div>
      <div
        data-slot="collapsible"
        data-state="closed"
      >
        <button
          aria-controls="radix-:r12:"
          aria-expanded="false"
          class="inline-flex items-center whitespace-nowrap font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:bg-accent hover:text-accent-foreground dark:hover:bg-accent/50 h-8 rounded-md gap-1.5 has-[>svg]:px-2.5 w-full justify-between p-2 text-xs"
          data-slot="collapsible-trigger"
          data-state="closed"
          type="button"
        >
          Technical Details
          <svg
            class="lucide lucide-chevron-down h-3 w-3"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m6 9 6 6 6-6"
            />
          </svg>
        </button>
        <div
          class="mt-2"
          data-slot="collapsible-content"
          data-state="closed"
          hidden=""
          id="radix-:r12:"
          style=""
        />
      </div>
      <div
        class="pt-2 border-t border-gray-200 text-xs text-gray-600"
      >
        <div
          class="flex justify-between items-center"
        >
          <span>
            Error ID: 
            mdthyic6
          </span>
          <span>
            8/1/2025, 5:11:03 PM
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ActivityErrorBoundary > Snapshot tests > should match snapshot for normal state 1`] = `
<div>
  Normal content
</div>
`;
